### ISSUE TITLE
1.21.1 - Can't start project: "the configured database type does not match the current actual database"

### ISSUE THREAD

**Issue by garvinhicking** [drud/ddev#4129]
TL;DR: Fix in https://github.com/drud/ddev/issues/4129#issuecomment-1220070181

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Run a Diagnostic and Paste Link Here

(need to check)

### Current Behavior

ddev project does not start, shows error:

`Failed to start fe-kuerschnersmart: Unable to start project fe-kuerschnersmart because the configured database type does not match the current actual database. Please change your database type back to “” and start again`

### Expected Behavior

Project getting booted

### Steps To Reproduce

Environment: MacOS
Previous ddev: 1.18.2
Installed via homebrew

config.yaml of project:

```yaml
name: fe-kuerschnersmart
router_http_port: "80"
router_https_port: "443"
mailhog_port: "15007"
type: typo3
docroot: htdocs/web/
nodejs_version: 16
webserver_type: apache-fpm
xdebug_enabled: false
additional_hostnames:
  ['bundestag.fe-kuerschnersmart', 'bw.fe-kuerschnersmart', 'home.fe-kuerschnersmart', 'relaunch.fe-kuerschnersmart']
additional_fqdns: []
mariadb_version: "10.2"
php_version: "7.4"
#mysql_version: "" # was previously contained, but doesn't matter if we leave it in or out
use_dns_when_possible: true
mutagen_enabled: false
composer_version: "2"
web_environment:
- SURF_WORKSPACE=/var/www/html/deployment/workspace
- SURF_HOME=/var/www/html/deployment
- GIT_SSH_COMMAND=ssh -i /var/www/html/deployment/ssh/id_rsa
- TYPO3_CONTEXT=Development/Ddev
hooks:
  post-start:
  - exec: bash /var/www/html/.ddev/scripts/vagrant-compat.sh
  - exec: bash /var/www/html/.ddev/scripts/sanity-check.sh
  - exec: nvm install
  - exec: nvm use
disable_settings_management: true
webimage_extra_packages: [build-essential, joe]
```

### Anything else?

A colleage had ddev 1.18.2 and upgraded to ddev - expecting version 1.20 that we others had, but he got the new version.

Since then he cannot boot up the project that was working previously. The changelog contains:

`ddev start now checks on startup to see if the existing database binary information is compatible with the configured database version and fails if it is not the same.``

which I believe could be the culprit here. We didn't change the mariadb_version in the project. Without the project starting, we don't know how to inspect the actual db volume, or what an error could be. Especially the error message with an empty string ""  strikes our mind. We ruled out the empty 'mysql_version' yaml string being the culprit.

Any ideas how we can debug this further or what could be an issue? I don't want to upgrade my working environment to stumble into the same situation because of important client work, so I'm not sure if I could reproduce it on my machine, too.

I've asked for the `ddev debug` output, but that may take some time.

**Comment by rfay** [comment]
@garvinhicking I'd love to help you figure out what is wrong.  Apparently DDEV can't figure out what database type/version is actually in your project; in v1.21.1, DDEV tries to prevent the classic problem where people change config and get an obscure message about the DB not coming up, so it tries to figure out what's out there.

* If you could please try `ddev debug get-volume-db-version` that command tries to get the version that's actually in the database volume. I have to assume you'll have the same problem with that. 
* `ddev debug check-db-match` attempts to see if you have a matching version to your config. I would assume that will fail for the same reason.

But the bottom line is perhaps some info in the DB volume just isn't there that DDEV is expecting.

Could you please
* Share the non-comment part of your .ddev/config.yaml?
* After you do that, do a `ddev config --auto` and see if a refreshed config changes things?
* After `ddev config --auto` please share the non-comment part of your .ddev/config.yaml

In both of the following you must change the project name from "d9" to the name of your project:
* Assuming you were using mariadb or mysql, please show the output of `docker run -it --rm -v <project>-mariadb:/tmp/db busybox ls -l /tmp/db`
* Again same assumption, please show the output of `docker run -it --rm -v <project>-mariadb:/tmp/db busybox cat /tmp/db/db_mariadb_version.txt`

**Comment by rfay** [comment]
Oh, I see you have the config.yaml above, and it's a very old config.yaml, using the pre-v1.19 syntax for database configuration, so it may be that the `ddev config --auto` will fix something.

**Comment by rfay** [comment]
And, of course, you can `ddev delete --omit-snapshot` as long as you have a DB dump or a snapshot and get past this problem. But if you could take the time to look through the ideas above before blowing away the evidence it would be appreciated.

**Comment by rfay** [comment]
I was able to get behavior something like yours by completely removing the DB config from the config.yaml, but it's not quite the same, and in my case it's because my existing db is mariadb:10.3 and the new default (that you get when not having any database config) is mariadb:10.4)

**Comment by rfay** [comment]
I was able to recreate this by creating a project with DDEV v1.18.2 and then upgrading. It seems that back in v1.18, the marker on the database was not the same.

```
docker run -it --rm -v d9-mariadb:/tmp/db busybox cat /tmp/db/db_mariadb_version.txt
10.3
```

I think you will be able to fix your particular project with:
```
docker run -it --rm -v fe-kuerschnersmart-mariadb:/tmp/db busybox sh -c 'echo "mariadb_10.2" > /tmp/db/db_mariadb_version.txt'
```

This uses the newer technique for marking the database volume's type.

**Comment by rfay** [comment]
I wrote a little script to do this for you, https://gist.github.com/rfay/27b12e49ceb1ed1e3959c603401b7010

You should be able to run it with `bash convert-old-db.sh fe-kuerschnersmart mariadb_10.2`

**Comment by adam-frontkom** [comment]
Hey rfay, thanks for providing the solution. I just happen to encouter it in one of my projects afer updating to the newest version of ddev.

Running the following command fixed it.
`docker run -it --rm -v projectname-mariadb:/tmp/db busybox sh -c 'echo "mariadb_10.3" > /tmp/db/db_mariadb_version.txt'`

**Comment by garvinhicking** [comment]
@rfay You're a superstar. Thanks so much for taking the time (which surely wasn't just a few minutes) to reproduce and investigate this. Due to CEST I was soundly asleep so couldn't join in, sorry.

Your findings make total sense, and I am absolutely sure they'll fix our issue. Due to client work on our machines we'll need to find a timeslot to make this change, for now we just downgraded ddev. We should be able to do this later next week and report back.

Is it likely that this issue will be addressed with an upcoming release, so that maybe the version inspection takes these migrated instances into consideration?

**Comment by rfay** [comment]
I will definitely see what can be done about this. v1.19 was a big thing in the db world because postgres was introduced, and that's where this changed. I think there was an upgrade then, but perhaps this check is running before the upgrade.

**Comment by rfay** [comment]
It turns out that the upgrade of that marker was definitely done by DDEV v1.19+:

https://github.com/drud/ddev/blob/49f6105d6eca041dc0e66bdbb212d51986ce7abd/containers/ddev-dbserver/files/docker-entrypoint.sh#L105-L110

The problem is that it's done at startup time, and the new do-you-have-the-right-version check is done before the database container gets run... to prevent obscure problems that people complained about.

So this but likely affects only people who upgraded a project directly from DDEV v1.18.* or earlier to v1.21.1.

I think it's reasonable to add a little code in the safety check that will allow these older versions to start up (and thus get the marker upgraded). The code would go on the golang side, and accomplish the same basic check that the bash container startup script is doing.

**Comment by rfay** [comment]
#4132 should sort this out for upcoming DDEV releases, or when it gets pulled, for people using HEAD version.

**Comment by jokumer** [comment]
Tested, using HEAD version - works. Thx

**Comment by briezler** [comment]
@rfay You are the best, and fastest. Thx works like a charme.
